fclient/performjoin.go

```diff
@@ -0,0 +1,238 @@
+	// request.
+	respMakeJoin, err := fedClient.MakeJoin(
+		ctx,
+		origin,
+		input.ServerName,
+		input.RoomID,
+		input.UserID.Raw(),
+	)
+	if err != nil {
+		// TODO: Check if the user was not allowed to join the room.
+		callbacks.FederationFailure(input.ServerName)
+		return nil, nil, fmt.Errorf("r.federation.MakeJoin: %w", err)
+	}
+	callbacks.FederationSuccess(input.ServerName)
+
+	// Set all the fields to be what they should be, this should be a no-op
+	// but it's possible that the remote server returned us something "odd"
+	stateKey := input.UserID.Raw()
+	respMakeJoin.JoinEvent.Type = spec.MRoomMember
+	respMakeJoin.JoinEvent.Sender = input.UserID.Raw()
+	respMakeJoin.JoinEvent.StateKey = &stateKey
+	respMakeJoin.JoinEvent.RoomID = input.RoomID
+	respMakeJoin.JoinEvent.Redacts = ""
+	if input.Content == nil {
+		input.Content = map[string]interface{}{}
+	}
+	_ = json.Unmarshal(respMakeJoin.JoinEvent.Content, &input.Content)
+	input.Content["membership"] = spec.Join
+	if err = respMakeJoin.JoinEvent.SetContent(input.Content); err != nil {
+		return nil, nil, fmt.Errorf("respMakeJoin.JoinEvent.SetContent: %w", err)
+	}
+	if err = respMakeJoin.JoinEvent.SetUnsigned(struct{}{}); err != nil {
+		return nil, nil, fmt.Errorf("respMakeJoin.JoinEvent.SetUnsigned: %w", err)
+	}
+
+	// Work out if we support the room version that has been supplied in
+	// the make_join response.
+	// "If not provided, the room version is assumed to be either "1" or "2"."
+	// https://matrix.org/docs/spec/server_server/unstable#get-matrix-federation-v1-make-join-roomid-userid
+	if respMakeJoin.RoomVersion == "" {
+		respMakeJoin.RoomVersion = setDefaultRoomVersionFromJoinEvent(respMakeJoin.JoinEvent)
+	}
+	verImpl, err := gomatrixserverlib.GetRoomVersion(respMakeJoin.RoomVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Build the join event.
+	event, err := respMakeJoin.JoinEvent.Build(
+		time.Now(),
+		origin,
+		input.KeyID,
+		input.PrivateKey,
+		respMakeJoin.RoomVersion,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("respMakeJoin.JoinEvent.Build: %w", err)
+	}
+
+	var respState gomatrixserverlib.StateResponse
+	// Try to perform a send_join using the newly built event.
+	respSendJoin, err := fedClient.SendJoin(
+		context.Background(),
+		origin,
+		input.ServerName,
+		event,
+	)
+	if err != nil {
+		callbacks.FederationFailure(input.ServerName)
+		return nil, nil, fmt.Errorf("r.federation.SendJoin: %w", err)
+	}
+	callbacks.FederationSuccess(input.ServerName)
+
+	// If the remote server returned an event in the "event" key of
+	// the send_join request then we should use that instead. It may
+	// contain signatures that we don't know about.
+	if len(respSendJoin.Event) > 0 {
+		var remoteEvent *gomatrixserverlib.Event
+		remoteEvent, err = verImpl.NewEventFromUntrustedJSON(respSendJoin.Event)
+		if err == nil && isWellFormedMembershipEvent(
+			remoteEvent, input.RoomID, input.UserID,
+		) {
+			event = remoteEvent
+		}
+	}
```

Comment:
We seemingly continue blindly if the event is invalid?

Reply:
I don't know what the logic should be if you receive a bad event inside a send_join response.
Synapse just accepts the event as is and doesn't seem to do any checks against it. So we differ from them in that regard. 

At this point in the join dance, the resident server has already accepted that event into it's room graph and propagated it out to the other servers in the room. So for us to reject it leaves us in an awkward situation.